src/python/pants/core/util_rules/environments.py

```diff
@@ -363,5 +319,130 @@ def extract_process_config_from_environment(tgt: EnvironmentTarget) -> ProcessCo
     return ProcessConfigFromEnvironment(docker_image=docker_image)
 
 
+logger = logging.getLogger(__name__)
+
+
+class EnvironmentSensitiveOptionFieldMixin(ABC):
+    subsystem: type[Subsystem]
+    option_name: str
+
+
+_rules_for_subsystems: dict[type[Subsystem], set[UnionRule]] = {}
+
+
+def add_option_fields_for(subsystem: type[Subsystem]) -> Iterable[UnionRule]:
+    """Register environment fields for the `environment_sensitive` options of `subsystem`
+
+    This should be called in the `rules()` method in the file where `subsystem` is defined. It will
+    register the relevant fields under the `local_environment` and `docker_environment` targets.
+    """
+    if subsystem in _rules_for_subsystems:
+        return _rules_for_subsystems[subsystem]
+
+    field_rules: set[UnionRule] = set()
+
+    for option_attrname, option in _collect_options_info_extended(subsystem):
+        if option.flag_options["environment_sensitive"]:
+            field_rules.update(_add_option_field_for(subsystem, option, option_attrname))
+
+    _rules_for_subsystems[subsystem] = field_rules
+    return field_rules
+
+
+def _add_option_field_for(
+    subsystem_t: type[Subsystem], option: OptionsInfo, attrname: str
+) -> Iterable[UnionRule]:
+    option_type: type = option.flag_options["type"]
+    scope = getattr(subsystem_t, "options_scope")
+
+    # If you get a `KeyError` due to a missing value in `map` or `list_map`, it means there
+    # isn't a mapping for a given option type added at the bottom of this file. Add it there
+    # (note that currently it's not immediately clear how to support `Enum` options)
+
+    ofm: _OptionFieldMap
+    if option_type != list:
+        ofm = _OptionFieldMap.map[option_type]
+    else:
+        member_type = option.flag_options["member_type"]
+        ofm = _OptionFieldMap.list_map[member_type]
+
+    # The below class will never be used for static type checking outside of this function.
+    # so it's reasonably safe to use `ignore[name-defined]`. Ensure that all this remains valid
+    # if `_OptionFieldMap` is ever modified.
+    class OptionField(ofm.field_type, EnvironmentSensitiveOptionFieldMixin):  # type: ignore[name-defined]
+        # TODO: use envvar-like normalization logic here
+        alias = f"{scope}_{option.flag_names[0][2:]}".replace("-", "_")
+        required = False
+        value: ofm.field_value_type  # type: ignore[name-defined]
+        help = option.flag_options["help"] or ""
+        subsystem = subsystem_t
+        option_name = attrname
+
+    # If we ever add a third environment type, this would be a good thing to genericise.
```

Comment:
It's probably fine to keep as is. We will add a `remote_execution_environment` in the future, but it will still have tight locality in this file and imo isn't a big deal to add a 3rd target here. There's already a lot of magic in this file, so keeping it simple is probably better